samples/net/latmon/src/main.c

```diff
@@ -0,0 +1,287 @@
+
+#if defined(LATMON_LOOPBACK_DBG)
+	/*
+	 * Connect GPIO pins in loopback mode for validation (tx to ack)
+	 * On FRDM_K64F, Latmus will show around 3.2 usec of latency.
+	 */
+	ret = gpio_pin_interrupt_configure_dt(&ack, GPIO_INT_EDGE_FALLING);
+#else
+	ret = gpio_pin_interrupt_configure_dt(&ack, GPIO_INT_EDGE_RISING);
+#endif
+	if (ret < 0) {
+		LOG_ERR("failed configuring ack pin interrupt");
+		return ret;
+	}
+
+	gpio_init_callback(&gpio_cb, gpio_ack_handler, BIT(ack.pin));
+
+	ret = gpio_add_callback_dt(&ack, &gpio_cb);
+	if (ret < 0) {
+		LOG_ERR("failed adding ack pin callback");
+		return ret;
+	}
+
+	return ret;
+}
+
+static struct in_addr *get_dhcp_address(void)
+{
+	struct net_if_config *cf = &net_if_get_default()->config;
+	static struct in_addr *ip;
+
+	for (int n = 0; !ip && n < NET_IF_MAX_IPV4_ADDR; n++) {
+		if (cf->ip.ipv4->unicast[n].ipv4.addr_type == NET_ADDR_DHCP) {
+			ip = &cf->ip.ipv4->unicast[n].ipv4.address.in_addr;
+		}
+	}
+
+	return ip;
+}
+
+static void blink(void*, void*, void*)
+{
+	const struct gpio_dt_spec led_run =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led0), gpios, {0});
+	const struct gpio_dt_spec led_wait =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led1), gpios, {0});
+	const struct gpio_dt_spec led_dhcp =
+		GPIO_DT_SPEC_GET_OR(DT_ALIAS(led2), gpios, {0});
+	const struct gpio_dt_spec *led = &led_dhcp, *tmp = NULL;
+	uint32_t period = LED_DHCP_PERIOD;
+
+	if (gpio_is_ready_dt(&led_run)) {
+		gpio_pin_configure_dt(&led_run, GPIO_OUTPUT_INACTIVE);
+	}
+
+	if (gpio_is_ready_dt(&led_wait)) {
+		gpio_pin_configure_dt(&led_wait, GPIO_OUTPUT_INACTIVE);
+	}
+
+	if (gpio_is_ready_dt(&led_dhcp)) {
+		gpio_pin_configure_dt(&led_dhcp, GPIO_OUTPUT_INACTIVE);
+	}
+
+	for (;;) {
+		k_usleep(period);
+		if (get_dhcp_address()) {
+			led = net_latmon_running() ? &led_run : &led_wait;
+		}
+
+		if (tmp && led != tmp) {
+			gpio_pin_set_dt(tmp, 0);
+		}
+
+		if (!gpio_is_ready_dt(led)) {
+			continue;
+		}
+
+		if (led == &led_wait) {
+			period = LED_WAIT_PERIOD;
+		}
+
+		if (led == &led_run) {
+			period = LED_RUN_PERIOD;
+		}
+
+		gpio_pin_toggle_dt(led);
+		tmp = led;
+	}
+	gpio_pin_set_dt(led, 0);
+}
+
+static k_tid_t start_led_blinking_thread(struct k_thread *blink_thread,
+					k_thread_entry_t blink_thread_func)
+{
+	return k_thread_create(blink_thread, blink_stack, BLINK_STACK_SIZE,
+				(k_thread_entry_t)blink_thread_func,
+				NULL, NULL, NULL,
+				BLINK_THREAD_PRIORITY, 0, K_NO_WAIT);
+}
+
+static void dhcp_handler(struct net_mgmt_event_callback *cb,
+			uint32_t mgmt_event, struct net_if *iface)
+{
+	if (mgmt_event == NET_EVENT_IPV4_DHCP_BOUND) {
+		k_sem_give(&dhcp_done);
+	}
+}
+
+static struct in_addr *acquire_dhcp_address(struct net_mgmt_event_callback *cb)
+{
+	struct net_if *iface = net_if_get_default();
+
+	LOG_INF("DHCPv4: binding...");
+	net_mgmt_init_event_callback(cb, dhcp_handler,
+				     NET_EVENT_IPV4_DHCP_BOUND);
+	net_mgmt_add_event_callback(cb);
+	net_dhcpv4_start(iface);
+	k_sem_take(&dhcp_done, K_FOREVER);
```

Comment:
Instead of creating a semaphore and a handler, you could do here
```
struct net_if_dhcpv4 *dhcpv4;
size_t info_len;

ret = net_mgmt_event_wait(NET_EVENT_IPV4_DHCP_BOUND, NULL, NULL, &dhcpv4, &info_len, timeout);
```
which would wait until the desired event is returned. The bound IPv4 address is then found in the dhcpv4 struct.

Reply:
I modified it slightly to avoid a memcpy.